`@_silgen_name` vs `@_cdecl`

Currently, I use `@_silgen_name` on the following lines.

- https://github.com/kkebo/swift_os/blob/cce13c886ccf80cb3a9b1903ed154096238fddbc/Sources/Kernel/kernel.swift#L1
- https://github.com/kkebo/swift_os/blob/cce13c886ccf80cb3a9b1903ed154096238fddbc/Sources/Kernel/putchar.swift#L1

However, there are two similar attributes, [`@_silgen_name`](https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md#_silgen_nameraw-cname) and [`@_cdecl`](https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md#_cdeclcname). The former is using the Swift calling convention and the latter is using the C calling convention. I don't understand the differences between the two.

First, I have to understand the Swift calling convention.

https://github.com/apple/swift/blob/main/docs/ABI/CallingConvention.rst

The latest main snapshot toolchain forces to use `@_cdecl` instead of `@_silgen_name`.

```console
$ swift --version
Swift version 6.0-dev (LLVM 70f5988c1c84591, Swift 4f8181dde6a5a62)
Target: aarch64-unknown-linux-gnu
$ make run
swift build --triple aarch64-none-none-elf -c release -Xswiftc -Osize --experimental-lto-mode=full -Xswiftc -experimental-hermetic-seal-at-link
Building for production...
/home/kebo/swift_os/Sources/Kernel/putchar.swift:4:13: error: type mismatch of function 'putchar', declared as '@convention(thin) (Int32) -> Int32' but used in a swift module as '@convention(c) (Int32) -> Int32'
2 | @discardableResult
3 | @inlinable
4 | public func putchar(_ c: CInt) -> CInt {
  |             `- error: type mismatch of function 'putchar', declared as '@convention(thin) (Int32) -> Int32' but used in a swift module as '@convention(c) (Int32) -> Int32'
5 |     putchar(UInt8(c))
6 |     return c

make: *** [Makefile:24: swift] Error 1
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -7,14 +7,14 @@ on:
 jobs:
   build-linux:
     runs-on: ubuntu-latest
-    container: swiftlang/swift:nightly-6.0-jammy
+    container: swiftlang/swift:nightly-main-jammy
     steps:
       - uses: actions/checkout@v4
       - run: apt-get update && apt-get install --no-install-recommends -y make llvm-14
       - run: make OBJCOPY=llvm-objcopy-14
   lint:
     runs-on: ubuntu-latest
-    container: swiftlang/swift:nightly-6.0-jammy
+    container: swiftlang/swift:nightly-main-jammy
     steps:
       - uses: actions/checkout@v4
       - run: swift format lint -rs .

--- a/README.md
+++ b/README.md
@@ -19,12 +19,12 @@ An operating system written in Swift.
   - Fedora: `sudo dnf install qemu-system-aarch64-core`
   - Ubuntu: `sudo apt install qemu-system-arm`
   - macOS: `brew install qemu`
-- Swift 6.0 (because Embedded Swift is used)
+- Swift trunk development (main) toolchain (because Embedded Swift is used)
   - On Linux, you can easily install the toolchain using [swiftly](https://swift-server.github.io/swiftly/).
 
     ```shell
     curl -L https://swift-server.github.io/swiftly/swiftly-install.sh | bash
-    swiftly install 6.0-snapshot
+    swiftly install main-snapshot
     ```
 
 ## Building

--- a/README.md
+++ b/README.md
@@ -33,13 +33,6 @@ An operating system written in Swift.
 make
 ```
 
-> [!WARNING]
-> If you used swiftly to install Swift, you have to modify `Makefile` until https://github.com/swift-server/swiftly/issues/92 is resolved.
->
-> ```makefile
-> SWIFT := swift-legacy-driver
-> ```
-
 ## Cleaning build outputs
 
 ```shell

--- a/Sources/Kernel/kernel.swift
+++ b/Sources/Kernel/kernel.swift
@@ -1,4 +1,4 @@
-@_silgen_name("kmain")
+@_cdecl("kmain")
 public func kmain() {
     initUART()
 

--- a/Sources/Kernel/putchar.swift
+++ b/Sources/Kernel/putchar.swift
@@ -1,4 +1,4 @@
-@_silgen_name("putchar")
+@_cdecl("putchar")
 @discardableResult
 @inlinable
 public func putchar(_ c: CInt) -> CInt {

